crates/ra_hir_def/src/import_map.rs

```diff
@@ -0,0 +1,323 @@
+                } else {
+                    Some((name, per_ns))
+                }
+            });
+
+            for (name, per_ns) in visible_items {
+                let mk_path = || {
+                    let mut path = mod_path.clone();
+                    path.segments.push(name.clone());
+                    path
+                };
+
+                for item in per_ns.iter_items() {
+                    let path = mk_path();
+                    match import_map.entry(item) {
+                        Entry::Vacant(entry) => {
+                            entry.insert(path);
+                        }
+                        Entry::Occupied(mut entry) => {
+                            // If the new path is shorter, prefer that one.
+                            if path.len() < entry.get().len() {
+                                *entry.get_mut() = path;
+                            } else {
+                                continue;
+                            }
+                        }
+                    }
+
+                    // If we've just added a path to a module, descend into it.
+                    if let Some(ModuleDefId::ModuleId(mod_id)) = item.as_module_def_id() {
+                        worklist.push((mod_id, mk_path()));
+                    }
+                }
+            }
+        }
+
+        Arc::new(Self { map: import_map })
+    }
+
+    /// Returns the `ModPath` needed to import/mention `item`, relative to this crate's root.
+    pub fn path_of(&self, item: ItemInNs) -> Option<&ModPath> {
+        self.map.get(&item)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::test_db::TestDB;
+    use insta::assert_snapshot;
+    use ra_db::fixture::WithFixture;
+    use ra_db::SourceDatabase;
+
+    fn import_map(ra_fixture: &str) -> String {
+        let db = TestDB::with_files(ra_fixture);
+        let crate_graph = db.crate_graph();
+
+        let import_maps: Vec<_> = crate_graph
+            .iter()
+            .filter_map(|krate| {
+                let cdata = &crate_graph[krate];
+                let name = cdata.display_name.as_ref()?;
+
+                let map = db.import_map(krate);
+
+                let mut importable_paths: Vec<_> = map
+                    .map
+                    .iter()
+                    .map(|(item, modpath)| {
+                        let ns = match item {
+                            ItemInNs::Types(_) => "t",
+                            ItemInNs::Values(_) => "v",
+                            ItemInNs::Macros(_) => "m",
+                        };
+                        format!("- {} ({})", modpath, ns)
+                    })
+                    .collect();
+
+                importable_paths.sort();
+                let importable_paths = importable_paths.join("\n");
+
+                Some(format!("{}:\n{}", name, importable_paths))
+            })
+            .collect();
+
+        import_maps.join("\n")
+    }
```

Comment:
Maybe pull this into `Display` for `ImportMap`? It looks like it might be useful to take a look at this repr when doing printf debugging

Reply:
I decided to implement `Debug` instead, but yeah this makes sense